source/base/ThreadPool.ooc

```diff
@@ -1,111 +1,203 @@
 use ooc-collections
+use ooc-base
 
 import structs/LinkedList
 import threading/Thread
+import threading/native/ConditionUnix
+import os/Time
 
-ThreadJob: class {
-	_body: Func
-	init: func (=_body)
-	_execute: func { this _body() }
-	run: virtual func {
-		this _execute()
-		this free()
-	}
+_TaskState: enum {
+	Unfinished
+	Finished
+	Cancelled
 }
 
-SynchronizedThreadJob: class extends ThreadJob {
-	_finishedCondition := WaitCondition new()
-	_mutex := Mutex new()
-	_finished := false
-	finished ::= this _finished
-	_freeOnCompletion := false
-	init: func (body: Func) { super(body) }
+_Task: class {
+	_mutex: Mutex
+	_action: Func
+	_state : _TaskState
+	_freeDirectly: Bool
+	init: func (=_action, =_mutex) {
+		_state = _TaskState Unfinished
+		_mutex = Mutex new()
```

Comment:
Okay, I'm starting to see what's going on here. Still, if you create a new `Mutex` here in the constructor and don't destroy it later, it will leak.